### PR TITLE
Some minor cpu info box enhancements

### PR DESCRIPTION
**Pull request by ThunderCls**
- Added quick instruction references view in cpu extra info widget
- Added menu entry to modify memory values in cpu extra info widget
- Added menu entry to show current instruction references from cpu extra
info widget
- Added menu entry to copy file offset in disassembly and dump widget

---
This change is [Reviewable](https://reviewable.io/reviews/x64dbg/x64dbg/1572)

### CODE REVIEW

File: src/gui/Src/BasicView/HexDump.cpp [hunk 1]
```diff
@@ -317,6 +317,12 @@ void HexDump::setupCopyMenu()
     connect(mCopyRva, SIGNAL(triggered()), this, SLOT(copyRvaSlot()));
     mCopyRva->setShortcutContext(Qt::WidgetShortcut);
     addAction(mCopyRva);
+
+    // Copy -> File Offset
+    mCopyFileOffset = new QAction(DIcon("fileoffset.png"), "&File Offset", this);
```
Comment (torusrxxx) [inline src/gui/Src/BasicView/HexDump.cpp]: Please use `tr()` function whenever you want to show a string.

File: src/gui/Src/Gui/CPUDisassembly.cpp [hunk 2]
```diff
@@ -1718,6 +1719,27 @@ void CPUDisassembly::copyRvaSlot()
...
+    {
+        if(i)
+            clipboard += "\r\n";
+        duint addr = rvaToVa(inst.rva);
+        duint offset = DbgFunctions()->VaToFileOffset(addr);
+        if(offset)
+            clipboard += ToHexString(offset);
+        else
+        {
+            SimpleWarningBox(this, tr("Error!"), tr("Selection not in a file..."));
```
Comment (torusrxxx) [inline src/gui/Src/Gui/CPUDisassembly.cpp]: Please select carefully between `SimpleWarningBox` and `SimpleErrorBox`

File: src/gui/Src/Gui/CPUInfoBox.cpp [hunk 3]
```diff
@@ -244,6 +267,52 @@ void CPUInfoBox::disasmSelectionChanged(dsint parVA)
...
+        for(duint i = 0; i < xrefInfo.refcount; i++)
+        {
+            if(t != xrefInfo.references[i].type)
+            {
+                if(t != -1)
+                    output += ", ";
+                switch(xrefInfo.references[i].type)
+                {
+                    case XREF_JMP:
+                        output += "Jump from ";
```
Comment (torusrxxx) [inline src/gui/Src/Gui/CPUInfoBox.cpp]: Use `tr()` function here

File: src/gui/Src/Gui/CPUInfoBox.cpp [hunk 4]
```diff
@@ -244,6 +267,52 @@ void CPUInfoBox::disasmSelectionChanged(dsint parVA)
     if(getInfoLine(0) == getInfoLine(1)) //check for duplicate info line
         setInfoLine(1, "");
 
+    // check references details
+    // code extracted from ExtraInfo plugin by torusxxx
+    XREF_INFO xrefInfo;
+    DbgXrefGet(parVA, &xrefInfo);
```
Comment (torusrxxx) [inline src/gui/Src/Gui/CPUInfoBox.cpp]: The usage of `DbgXrefGet` is incorrect and will cause memory leak. The ExtraInfo plugin has been updated to fix it. Please use the newer version of ExtraInfo.

File: src/gui/Src/Gui/CPUInfoBox.cpp [hunk 5]
```diff
@@ -316,6 +385,83 @@ void CPUInfoBox::followActionSlot()
...
+    emit displayReferencesWidget();
+}
+
+void CPUInfoBox::addModifyValueMenuItem(QMenu* menu, QString name, duint value)
+{
+    foreach(QAction* action, menu->actions()) //check for duplicate action
+        if(action->text() == name)
+            return;
+    QAction* newAction = new QAction(name, menu);
+    newAction->setFont(QFont("Courier New", 8));
```
Comment (torusrxxx) [inline src/gui/Src/Gui/CPUInfoBox.cpp]: We should use user-configurable font whenever possible.

**Comment by ThunderCls** [comment]
Thanks for your comments @torusrxxx I'll update my pr as soon as possible

File: src/gui/Src/BasicView/HexDump.h [hunk 6]
```diff
@@ -217,6 +218,7 @@ public slots:
     QAction* mCopyAddress;
     QAction* mCopyRva;
     QAction* mCopySelection;
+    QAction* mCopyFileOffset;
```
Comment (mrexodia) [inline src/gui/Src/BasicView/HexDump.h]: This action should be created with `makeShortcutAction` in `CPUDump.cpp` not over here.

File: src/gui/Src/Gui/CPUInfoBox.cpp [hunk 7]
```diff
@@ -244,6 +267,52 @@ void CPUInfoBox::disasmSelectionChanged(dsint parVA)
...
 
+    // check references details
+    // code extracted from ExtraInfo plugin by torusxxx
+    XREF_INFO xrefInfo;
+    DbgXrefGet(parVA, &xrefInfo);
+    if(xrefInfo.refcount > 0)
+    {
+        QString output;
+        char clabel[MAX_LABEL_SIZE] = "";
+        std::qsort(xrefInfo.references, xrefInfo.refcount, sizeof(xrefInfo.references[0]), &compareFunc);
```
Comment (mrexodia) [inline src/gui/Src/Gui/CPUInfoBox.cpp]: Don't use this. use `std::sort` with a lambda.

File: src/gui/Src/Gui/CPUInfoBox.cpp [hunk 8]
```diff
@@ -371,7 +517,7 @@ void CPUInfoBox::setupFollowMenu(QMenu* menu, duint wVA)
                     addFollowMenuItem(menu, tr("&Constant: ") + constant, arg.constant);
             }
             if(DbgMemIsValidReadPtr(arg.memvalue))
-                addFollowMenuItem(menu, tr("&Value: ") + segment + "[" + QString(arg.mnemonic) + "]", arg.memvalue);
+                addFollowMenuItem(menu, tr("&Value: ") + segment + "[" + QString(arg.mnemonic).toUpper().trimmed() + "]", arg.memvalue);
```
Comment (mrexodia) [inline src/gui/Src/Gui/CPUInfoBox.cpp]: This shouldn't be uppercase.

File: src/gui/Src/Gui/CPUInfoBox.cpp [hunk 9]
```diff
@@ -316,6 +385,83 @@ void CPUInfoBox::followActionSlot()
...
+            if(DbgMemIsValidReadPtr(arg.value))
+                addModifyValueMenuItem(menu, tr("&Address: ") + segment + QString(arg.mnemonic).toUpper().trimmed(), arg.value);
+            if(arg.value != arg.constant)
+            {
+                QString constant = QString("%1").arg(ToHexString(arg.constant));
+                if(DbgMemIsValidReadPtr(arg.constant))
+                    addModifyValueMenuItem(menu, tr("&Constant: ") + constant, arg.constant);
+            }
+            if(DbgMemIsValidReadPtr(arg.memvalue))
+                addModifyValueMenuItem(menu, tr("&Value: ") + segment + "[" + QString(arg.mnemonic).toUpper().trimmed() + "]", arg.memvalue);
```
Comment (mrexodia) [inline src/gui/Src/Gui/CPUInfoBox.cpp]: These shouldn't be uppercase either.

File: src/gui/Src/Gui/MainWindow.cpp [hunk 10]
```diff
@@ -338,6 +338,8 @@ MainWindow::MainWindow(QWidget* parent)
 
     connect(mCpuWidget->getStackWidget(), SIGNAL(displayReferencesWidget()), this, SLOT(displayReferencesWidget()));
 
+    connect((QObject*)mCpuWidget->getInfoBoxWidget(), SIGNAL(displayReferencesWidget()), this, SLOT(displayReferencesWidget()));
```
Comment (mrexodia) [inline src/gui/Src/Gui/MainWindow.cpp]: The cast to `QObject*` is not needed and also dangerous.

Comment (ThunderCls) [inline src/gui/Src/Gui/MainWindow.cpp]: The compiler complains about this being without a cast
image

**Comment by mrexodia** [comment]
Then it means you cannot do it. Casting doesn't solve anything, it will
just be undefined behavior.

File: src/gui/Src/Gui/MainWindow.cpp (hunk 10 above)
Comment (torusrxxx) [inline src/gui/Src/Gui/MainWindow.cpp]: It is probably you have not included the header for info box yet.

**Comment by ThunderCls** [comment]
Fixes done!

**Comment by ThunderCls** [comment]
Ok this PR went all messed up, I'm making a new push...sorry about that

**Pull request closed without merging**